src/compute-client/src/metrics.rs

```diff
@@ -454,70 +473,157 @@ impl StatsCollector<ProtoComputeCommand, ProtoComputeResponse> for ReplicaMetric
+    wallclock_lag_seconds_min: Gauge,
+    /// Gauge tracking maximum dataflow wallclock lag.
+    wallclock_lag_seconds_max: Gauge,
+    /// Counter tracking the total sum of dataflow wallclock lag.
+    wallclock_lag_seconds_sum: Counter,
+    /// Counter tracking the total count of dataflow wallclock lag measurements.
+    wallclock_lag_seconds_count: IntCounter,
+
+    /// State maintaining minimum and maximum wallclock lag.
+    wallclock_lag_minmax: MinMax<f32>,
 }
 
 impl ReplicaCollectionMetrics {
     pub fn observe_wallclock_lag(&mut self, lag: Duration) {
-        const REFRESH_INTERVAL: Duration = Duration::from_secs(60);
+        let lag_secs = lag.as_secs_f32();
 
-        self.wallclock_lag.add(lag);
+        self.wallclock_lag_minmax.add_sample(lag_secs);
 
-        if self.wallclock_lag.last_reset.elapsed() >= REFRESH_INTERVAL {
-            self.wallclock_lag_max_seconds
-                .set(self.wallclock_lag.max.as_secs_f64());
-            self.wallclock_lag_min_seconds
-                .set(self.wallclock_lag.min.as_secs_f64());
-            self.wallclock_lag_avg_seconds
-                .set(self.wallclock_lag.avg_secs());
+        let (&min, &max) = self
+            .wallclock_lag_minmax
+            .get()
+            .expect("just added a sample");
 
-            self.wallclock_lag.reset();
-        }
+        self.wallclock_lag_seconds_min.set(min.into());
+        self.wallclock_lag_seconds_max.set(max.into());
+        self.wallclock_lag_seconds_sum.inc_by(lag_secs.into());
+        self.wallclock_lag_seconds_count.inc();
     }
 }
 
-/// Helper for producing per-minute min/max/avg aggregates of observed wallclock lag.
+/// Keeps track of a the minimum and maximum value over a fixed-size sliding window of samples.
+///
+/// Inspired by the [`movning_min_max`](https://crates.io/crates/moving_min_max) crate, but
+/// adapted to provide:
+///
+///  * both minimum and maximum at the same time
+///  * a fixed window size, rather than manual popping of samples
+///  * lower memory usage, due to merged push and pop stacks
 #[derive(Debug)]
-struct WallclockLag {
-    count: usize,
-    sum: Duration,
-    max: Duration,
-    min: Duration,
-    last_reset: Instant,
+struct MinMax<T> {
+    /// The push stack and the pop stack, merged into one allocation.
+    ///
+    /// The push stack is the first `push_stack_len` items, the pop stack is the rest.
+    /// The push stack grows to the left, the pop stack grows to the right.
+    ///
+    /// +--------------------------------+
+    /// | push stack --> | <-- pop stack |
+    /// +----------------^---------------^
+    ///           push_stack_len
+    ///
+    /// We keep both the push stack and the pop stack in the same `Vec`, to guarantee that we only
+    /// allocate `size_of::<T>() * 3 * window_size` bytes. If we kept two `Vec`s we'd need twice
+    /// that much.
+    stacks: Vec<(T, T, T)>,
+    /// The length of the push stack.
+    push_stack_len: usize,
 }
```

Comment:
This is so so cool. I had to stare at it for a while to get an intuition for how it works, but I get it now. You fill up the whole vector up to its capacity with the push stack. Then when you hit the capacity, you flip it. Then you fill it up again one at a time, where each new entry on the push stack overwrites the previous entry on the pop stack. And when you get to the end, you flip it again, and start over. Genius!

Reply:
Indeed! I'll explain this better when moving this to ore.